File: Java/347.Top_K_Frequent_Elements.java
import java.util.HashMap;
import java.util.PriorityQueue;

class Solution {
    public List<Integer> topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> freq_map = new HashMap<>();
        for (int num: nums) {
            if (freq_map.containsKey(num)) {
                freq_map.put(num, freq_map.get(num) + 1);
            } else {
                freq_map.put(num, 1);
            }
        }

        PriorityQueue<Integer> heap = new PriorityQueue<>((a,b) -> freq_map.get(a) - freq_map.get(b));
        int counter = 0;
        for (int num : freq_map.keySet()) {
            if (counter < k) {
                heap.add(num);
                counter++;
            } else {
                int smallest_in_heap = heap.peek();
                if (freq_map.get(num) > freq_map.get(smallest_in_heap)) {
                    heap.poll();
                    heap.add(num);
                }
            }
        }
        List<Integer> ans = new ArrayList<>();
        for (int i = 0; i < heap.size();) {
            ans.add(heap.poll());
        }

        return ans;
    }
}

File: Python3/347.Top_K_Frequent_Elements.py
import heapq
from collections import defaultdict

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq_map = defaultdict(int)
        for num in nums:
            freq_map[num] += 1
        heap = []
        counter = 0
        for num, freq in freq_map.items():
            if (counter < k):
                heapq.heappush(heap, (freq, num))
                counter += 1
                continue
            smallest_in_heap = heapq.heappop(heap)
            if (freq > smallest_in_heap[0]):
                heapq.heappush(heap, (freq, num))
                continue
            heapq.heappush(heap, smallest_in_heap)
            
        
        for i in range(len(heap)):
            heap[i] = heap[i][1]

        return heap

        

        